Feat/432 add new filters to provide more granular control

bjorn2404 opened the pull request:
### Description of the Change

Added filters to provide more granular control over new OpenAI features as outlined in https://github.com/10up/classifai/issues/432

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #432

### How to test the Change



### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Fixed comment typo in includes/Classifai/Providers/Provider.php
> Added `classifai_chatgpt_allowed_image_roles` filter to allow ChatGPT image role settings to be overridden
> Added `classifai_openai_dalle_allowed_image_roles` filter to allow DALL·E image role settings to be overridden
> Added `classifai_openai_dalle_user_can_{$feature}` filter to allow granular control for DALL·E image generation
> Added `classifai_chatgpt_user_can_{$feature}` filter to allow granular control for ChatGPT title and excerpt generation


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @bjorn2404 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

Commit:
Fix typo

Commit:
Add image role setting filters to ChatGPT and DallE image processing classes

bjorn2404 commented:
@dkotter Wanted to double check on your intention from https://github.com/10up/classifai/issues/432. You had mentioned to update `ChatGPT.php` and `DallE.php`, which appears to just output the settings in the admin. If more of an override filter is wanted I was thinking of just adding one to the bottom of `generate_image_permissions_check()` within `includes/Classifai/Services/ImageProcessing.php`. Apologies if you had something else in mind - first time looking at this plugin and just interpreting everything. Thanks!

dkotter commented:
@bjorn2404 Thanks for the questions!

> If more of an override filter is wanted I was thinking of just adding one to the bottom of generate_image_permissions_check() within includes/Classifai/Services/ImageProcessing.php

So this would work for image generation but only if that is triggered via the REST API. This should be true for most use cases, someone in theory could directly use the methods within the `DallE` class. I personally would recommend adding a filter to the `generate_image_callback` within the `DallE` class instead, though I think some refactoring will be needed there to make this work.

In essence what I'm thinking is it's very likely that a client will want more fine-tune control over which users can use these features (since each request costs them money). As an example, they may have a single admin and two editors that should have access but none of the other admins or editors (or other account types). While we have settings in place to limit access based on role, that wouldn't help in this situation.

So having some sort of override filter that allows you to have that level of control could be very useful and is something we do in a few other places, like within the `OCR` class or `Read` class (though to be fair, these are more about allowing you to process items that you may not otherwise and less about user permissions).

I like the idea of introducing a new method to check if a user has access to the feature, that by default can have the checks we do right now (like checking the user role matches what is in the settings) but then has a filter around the final value, allowing someone total control on who can use the feature (this is just one potential approach though).

For `ChatGPT`, we have both a `generate_excerpt` and `generate_titles` methods that would be nice to support with this feature, as well as the `generate_image_callback` for `DallE`.

Commit:
Merge branch 'develop' into feat/432-Add-new-filters-to-provide-more-granular-control

Commit:
Add user_can_access method to DallE class

Commit:
Add user_can_access method with filtering to ChatGPT class

Commit:
Update user_can_access within DallE class so it can be used for additional features in the future if needed

Commit:
Pass specific role information along to the ChatGPT user_can_access filter

bjorn2404 commented:
@dkotter Thanks for the info! Does this look closer now to what you were thinking? I've added a `user_can_access` method to both classes where the setting name is passed through to the filter dynamically. Naming is maybe a little odd but I thought it would be nice to pass the setting name along as part of the feature filter name - ex: `classifai_chatgpt_user_can_enable_excerpt`.

Also, not sure if it's worth keeping the previous allowed_image_roles filters I added. Let me know if you think there would be of any value. Otherwise, I will update to remove them.

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -189,6 +226,19 @@ public function setup_fields_sections() {
+		 * Filter the allowed WordPress roles for ChatGTP
+		 *
+		 * @since x.x.x
+		 * @hook classifai_chatgpt_allowed_image_roles
```
I don't think this filter should have `image` in it, I'm assuming this was probably just a copy/paste error

phpbits replied:
@dkotter The fix for this is now ready for review. Thanks!

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -61,6 +61,43 @@ public function __construct( $service ) {
+		$access   = false;
+		$settings = $this->get_settings();
+
+		if (
+			! empty( $settings ) &&
+			isset( $settings['authenticated'] ) &&
+			false !== $settings['authenticated'] &&
```
I would suggest we remove the authentication checks here.

1. Seems to go against what this function says it does (checking if a user can access)
2. And we recently introduced an `is_configured` method that does this

phpbits replied:
@dkotter, The fix for this is now ready for review. Could you point me in the right direction for the `is_configured` 
 method? This needs to be added as well. Thanks!

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -61,6 +61,43 @@ public function __construct( $service ) {
+			! empty( $settings ) &&
+			isset( $settings['authenticated'] ) &&
+			false !== $settings['authenticated'] &&
+			isset( $settings[ $feature ] ) &&
+			'no' !== $settings[ $feature ]
```
Similar to the above, not sure I'd put this check here, was this checks if a feature is enabled, not if the user has access to the feature. I think I'd recommend removing this and instead adding our role checking in this method (which currently that hasn't been moved). So if we aren't authenticated or if the feature isn't turned on, no one will have access. But if both of those are true, we can then run this method to determine if an individual user should have access (by default, just checking if their role is allowed but developers can hook in and change access on a per user basis if needed).

phpbits replied:
@dkotter, The fix for this is now ready for review. 

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -87,7 +124,7 @@ public function enqueue_editor_assets() {
 
 		if (
 			( ! empty( $excerpt_roles ) && empty( array_diff( $user_roles, $excerpt_roles ) ) )
```
As mentioned above, I'd move this check within the new `user_can_access` method

phpbits replied:
@dkotter, The fix for this is now ready for review. 

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -87,7 +124,7 @@ public function enqueue_editor_assets() {
 		if (
 			( ! empty( $excerpt_roles ) && empty( array_diff( $user_roles, $excerpt_roles ) ) )
-			&& ( isset( $settings['enable_excerpt'] ) && 1 === (int) $settings['enable_excerpt'] )
+			&& $this->user_can_access( 'enable_excerpt', $excerpt_roles )
```
As mentioned above, I'd leave this enabled check alone rather than moving it into the new method

phpbits replied:
@dkotter, The fix for this is now ready for review. 

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -423,7 +473,7 @@ public function generate_excerpt( int $post_id = 0 ) {
 		// These checks (and the one above) happen in the REST permission_callback,
 		// but we run them again here in case this method is called directly.
-		if ( empty( $settings ) || ( isset( $settings['authenticated'] ) && false === $settings['authenticated'] ) || ( isset( $settings['enable_excerpt'] ) && 'no' === $settings['enable_excerpt'] ) ) {
+		if ( ! $this->user_can_access( 'enable_excerpt' ) ) {
```
As mentioned above, if we remove auth checks and enabled checks from the `user_can_access` method, I think these lines can stay as they are

phpbits replied:
@dkotter, The fix for this is now ready for review. 

cc @jeffpaul

dkotter commented on includes/Classifai/Providers/OpenAI/DallE.php:
```diff
@@ -52,22 +52,48 @@ public function __construct( $service ) {
 			&& ( ! empty( $roles ) && empty( array_diff( $user_roles, $roles ) ) )
-			&& ( isset( $settings['enable_image_gen'] ) && 1 === (int) $settings['enable_image_gen'] )
-		) {
+			&& ( isset( $settings[ $feature ] ) && 1 === (int) $settings[ $feature ] ) ) {
```
Similar here, I'd move this enabled check out of this method

phpbits replied:
@dkotter Could you point me in the right direction for the `is_configured` 
 method?

cc @jeffpaul

dkotter commented:
In addition to some of the above comments, we have user role checking that happens in our REST endpoints which will need to be handled (check out the `LanguageProcessing.php` and `ImageProcessing.php` Services)

phpbits committed:
Use classifai_chatgpt_allowed_roles filter

phpbits committed:
Remove authentication condition in user_can_access function

phpbits committed:
Fix user_can_access function for ChatGPT

dkotter committed:
Merge branch 'develop' into feat/432-Add-new-filters-to-provide-more-granular-control

dkotter committed:
Standardize on our enabled check. Ensure this check is used on the REST API side as well

dkotter committed:
Standardize on our enabled check. Ensure this check is used on the REST API side as well

dkotter committed:
Change filter names
